### DISCUSSION TITLE
"--concurrent" flag is not applied for failed helm releases

### DISCUSSION THREAD

**Discussion by ghost** [fluxcd/flux2#3832] [Q&A] [answered]
I have tried to increase number of helm releases helm controller processes concurrently by setting `--concurrent` flag. However, I have noticed strange behavior:
Concurrency is only increased when:

- new helm releases are created (in this case they are processed according to new concurrency value)
- helm releases that had READY status set to True are updated (in this case updated releases are processed according to new concurrency value)

However if helm releases that had READY status False (or were previously failed for some reason) are updated such releases are still reconciled only 4 at a time. `--concurrent` flag value is disregarded .

Is there a separate behavior or separate flag for failed release reconciliation parallelism ?

**Comment by kingdonb** [comment, marked as answer]
There is only one `--concurrent` flag, this sets the behavior of the Helm Controller with respect to how many resources Helm Controller can reconcile at the same time. Each resource is reconciled independently, the individual resources don't experience their own concurrency as an individual thing.

What might be happening that would give the appearance that some helm releases are blocking after failure, while others go through... is if the source controller hasn't also had its own concurrency setting bumped, that concurrency setting could potentially limit the number of HelmChart resources that Source Controller can make ready concurrently, which would in turn block HelmReleases.

How are the new helmreleases updated according to the new concurrency value, how do you observe that? The behavior you're describing doesn't sound correct.

There are some updates that Source Controller, Image Reflector, Kustomize Controller have gotten which haven't made it down to Helm Controller, Image Automation controller yet, where more granular status details are reported as conditions while the resource gets reconciled. When those changes land, it will become much easier to see where Helm Controller and Image Automation are stuck. Those changes should be coming soon, they have been in the works for a long while.

**Reply by ghost** [reply]
thanks for reply, I will try to bump up the concurrency setting in source controller too. Is it done via the same `--concurrent` flag ?

**Reply by ghost** [reply]
Thanks, setting concurrency to source controller did help. 
Do you happen to have any suggestions regarding another issue I'm having: https://github.com/fluxcd/flux2/discussions/3757
I always get `HelmChart 'NAME' is not ready` when new helm release is created.

**Reply by kingdonb** [reply]
In the bootstrap cheatsheet, we recommend bumping the concurrency in three controllers and show how to do that with a single patch:

https://fluxcd.io/flux/cheatsheets/bootstrap/#increase-the-number-of-workers

I will check out the link you sent and reply there 👍

**Reply by ghost** [reply]
Thanks a lot
